### DISCUSSION TITLE
compatibility of soft for arduino uno and at1616

### DISCUSSION THREAD

**Discussion by maulionc** [SpenceKonde/megaTinyCore#511] [Q&A]
hello 
I try to migrate a code for an arduino uno to an at1616. 
i've some space error like this 
![image](https://user-images.githubusercontent.com/89079859/131544388-8c177859-dcb6-41fe-9589-43ed979d8c82.png)
why megatinycore create this type of error ? 
Charles

**Comment by SpenceKonde** [comment]
Huh. I would expect "u8 does not name a type" 

The reason you get that error with megaTinyCore is because "u8" is a dumbass made-up type name. The only reason it works on the official AVR boards (it doesn't work on anything else, even official 'megaavr' `*` boards), is that the official avr board package's USBAPI.h, which because it's part of the core, is automatically included in all sketches even if compiling for a part without USB, contains the blasphemous line: `typedef unsigned char u8;`

`u8`  is wrong - the only place it is only bad practice instead of being objectively wrong is in code which is specific to classic AVR devices with native USB, and hence where dependance on USBAPI.h is normal and expected. If your sketch is declaring variables of type `u8`, it is correct that a compile error is generated. It should be replaced in your code with the c standard type name: `uint8_t` - or `unsigned char (also valid) `, or even (not recommended - not standard compliant) `byte` (byte is not standard C,  but at least it's part of the Arduino API so all cores are guaranteed to define it now and in the future.
If it is coming from a library, unless that library relates to native USB functionality on classic AVRs (ex, Arduinio Leonardo/Micro), it is a defect in the library that should be reported to the maintainer, because it won't work on any platform other than the official AVR boards. If the library does relate to native USB, the library is not compatible with megaTinyCore because none of these parts have native USB. 

`*` Speaking of dumbass made-up names, "megaavr", which is what Arduino calls the variation on the AVR architecture used on all post 2016 AVR devices (a convention that megaTinyCore must follow otherwise libraries will not work with it) is a name some guy at Arduino came up with, and which they decided to go with **despite the fact that Atmel already used that name** just with capital letters, to refer to something else! (arduino's megaavr is the same thing that I refer to as "modern AVR", after both Microchip support and the Microchip employee who I asked told me that there was no official name to differentiate between the modern and classic AVRs). In official Atmel/Microchip lingo, "megaAVR" refers to any AVR device with "ATmega" in the name. So the ATmega328p is a "megaAVR" according to Microchip, though not a "megaAVR 0-series" (like the ATmega4809 would be - see, the 4809 is both a "megaAVR" and a "megaaavr"), while in Arduino-land, the same part is **not** a "megaavr". The ATtiny3216 is not a "megaAVR" per Microchip (it's a "tinyAVR"), but it is a "megaavr"...
